lib/msf/core/rpc/v10/rpc_session.rb

```diff
@@ -279,6 +279,29 @@ def rpc_meterpreter_write( sid, data)
     { "result" => "success" }
   end
 
+  def rpc_meterpreter_execute(sid, data)
+    s = _valid_session(sid, "meterpreter")
+
+    s.single_session_mutex.synchronize {
```

Comment:
I'm a little wary of exposing a mutex to end-users of this API. I would prefer to add a convenience method to meterpreter sessions that handles this logic that this RPC then calls (it could possibly cover some of the other RPC use cases in this file as well).

Reply:
But I think it's a very good function when I want to wait for the command to return. Instead of risking multiple concurrent callers writing and reading the same Meterperter session can lead to
a conflict, where one caller gets the others output and vice versa.

Reply:
absolutely @akkuman it's a great idea, I think what @acammack-r7 is getting at (correct me if I've misunderstood you) is that it doesn't belong as part of the RPC api, not that the functionality shouldn't exist so we could look at moving this logic elsewhere